### PR TITLE
Add foreign keys to Ahoy models

### PR DESCRIPTION
**Pull request by rhymes**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [ ] Bug Fix
- [x] Optimization
- [ ] Documentation Update

## Description

As a follow up to https://github.com/forem/forem/pull/9565 I'm adding FKs to Ahoy::Models

By looking at the DB I found:

- 36 visits assigned to non existent users
- 1 event assigned to a non existent users
- couldn't run the query for messages as it times out, see:

```sql
select user_id from ahoy_messages where user_id is not null and user_id not in (select id from users);
```

## [optional] Are there any post deployment tasks we need to perform?

I added scripts to cleanup the data, please double check the logic is correct

### CODE REVIEW

File: lib/data_update_scripts/20200729120730_remove_orphaned_ahoy_events.rb [hunk 1]
```diff
@@ -1,11 +1,7 @@
 module DataUpdateScripts
   class RemoveOrphanedAhoyEvents
     def run
-      Ahoy::Event.find_each do |event|
-        next if event.visit.blank?
-
-        event.destroy
-      end
+      # Deleted the script as it was not necessary
```
Comment (rhymes) [inline lib/data_update_scripts/20200729120730_remove_orphaned_ahoy_events.rb:4]: this was removed as it contained the wrong logic

File: lib/data_update_scripts/20200805102636_remove_orphaned_ahoy_rows.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
+module DataUpdateScripts
+  class RemoveOrphanedAhoyRows
+    def run
+      user_ids = User.ids
+
+      # Delete all Ahoy::Events belonging to a user that does not exist anymore
+      Ahoy::Event.where.not(user_id: nil).where.not(user_id: user_ids).find_each(&:destroy)
+
+      # Delete all Ahoy::Messages belonging to a user that does not exist anymore
+      Ahoy::Message.where.not(user_id: nil).where.not(user_id: user_ids).find_each(&:destroy)
```
Comment (mstruve) [inline lib/data_update_scripts/20200805102636_remove_orphaned_ahoy_rows.rb:10]: There are 40255 Ahoy Messages that belong to missing users. I was able to run this in a console
```
ids = ActiveRecord::Base.connection.execute("select id from ahoy_messages where user_id is not null an
d user_id not in (select id from users);").map(&:itself);
```
However, using the user_ids with Rails notation like this in an array causes Postgres to search that whole array and even for Ahoy::Events, the query took 5+ seconds just for a count. I think to ensure this script completes we need to use straight SQL to get the job done. Then we can also ensure that our STATEMENT_TIMEOUT is bumped a little before we deploy to give the requests a little extra time to finish.

Comment (rhymes) [reply lib/data_update_scripts/20200805102636_remove_orphaned_ahoy_rows.rb:10]: Make sense! I'll rewrite them using SQL :-)

Comment (rhymes) [reply lib/data_update_scripts/20200805102636_remove_orphaned_ahoy_rows.rb]: Done! I wrote three SQL queries.

I wish PostgreSQL's SQL supported CTEs like this but alas it doesn't:

```sql
WITH user_ids AS (
  SELECT id from users
)

DELETE FROM ahoy_events
WHERE user_id IS NOT NULL
AND user_id NOT IN (SELECT ID FROM user_ids)

DELETE FROM ahoy_messages
WHERE user_id IS NOT NULL
AND user_id NOT IN (SELECT ID FROM user_ids)

DELETE FROM ahoy_visits
WHERE user_id IS NOT NULL
AND user_id NOT IN (SELECT ID FROM user_ids)
```

File: db/migrate/20200805102249_validate_add_missing_foreign_keys_to_ahoy_models.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+class ValidateAddMissingForeignKeysToAhoyModels < ActiveRecord::Migration[6.0]
+  def change
+    validate_foreign_key :ahoy_events, :ahoy_visits
+    validate_foreign_key :ahoy_events, :users
```
Comment (mstruve) [inline db/migrate/20200805102249_validate_add_missing_foreign_keys_to_ahoy_models.rb:4]: Since we have invalid data in the database, won't this fail? I think we may need to separate the validating into a separate PR.

Comment (rhymes) [reply db/migrate/20200805102249_validate_add_missing_foreign_keys_to_ahoy_models.rb:4]: Maybe we could create a PR that cleans the DB and then merge this one. What do you think?

Comment (mstruve) [reply db/migrate/20200805102249_validate_add_missing_foreign_keys_to_ahoy_models.rb:4]: That sounds good to me!

File: lib/data_update_scripts/20200805102636_remove_orphaned_ahoy_rows.rb (hunk 2 above)
Comment (citizen428) [reply lib/data_update_scripts/20200805102636_remove_orphaned_ahoy_rows.rb]: Not that it matters, but you should be able to do this with variable instead of a CTE:

```sql
DO $$
  DECLARE user_ids bigint[];
BEGIN
  user_ids := ARRAY(SELECT id from users);
  DELETE from ahoy_events
  WHERE user_id IS NOT NULL
  AND user_id != ALL(user_ids);
  
  ...
END $$;
```

Comment (rhymes) [reply lib/data_update_scripts/20200805102636_remove_orphaned_ahoy_rows.rb]: Oh nice, I didn't think of the `DO` block! TIL :D

**Comment by rhymes** [comment]
To be merged after https://github.com/forem/forem/pull/9710 is merged and deployed

File: db/migrate/20200805102249_validate_add_missing_foreign_keys_to_ahoy_models.rb (hunk 3 above)
Comment (maestromac) [reply db/migrate/20200805102249_validate_add_missing_foreign_keys_to_ahoy_models.rb:4]: This is good now
